costar_dataset/block_stacking_reader_torch.py

```diff
@@ -588,6 +589,74 @@ def transform_matrix_offset_center(matrix, x, y):
     transform_matrix = np.dot(np.dot(offset_matrix, matrix), reset_matrix)
     return transform_matrix
 
+def generate_training_data(video):
+        """Input data: numpy array of frames of a video
+           Output: classify how many frames there are between two random frames.
+        """
+        frames = np.empty((2, *video.shape[1:]))
+        label = np.zeros(6)
+
+        interval = random.randint(0, 5)
+        if interval == 0:
+            possible_frames_start = 0
+            possible_frames_end = 0
+        elif interval == 1:
+            possible_frames_start = 1
+            possible_frames_end = 1
+        elif interval == 2:
+            possible_frames_start = 2
+            possible_frames_end = 2
+        elif interval == 3:
+            possible_frames_start = 3
+            possible_frames_end = 4
+        elif interval == 4:
+            possible_frames_start = 5
+            possible_frames_end = 20
+        elif interval == 5:
+            possible_frames_start = 21
+            possible_frames_end = 150
+
+        first_frame_index = random.randint(0, video.shape[0] - possible_frames_end - 1)
+        second_frame_index = random.randint(possible_frames_start, possible_frames_end)
+
+        frames[0] = video[first_frame_index] / 255 # To normalize data
+        frames[1] = video[first_frame_index + second_frame_index] / 255
+        label[interval] = 1.
+        return (frames[0],frames[1], interval)
+
+def generate_crossmodal_training_data(video,joints, length = 10):
+    """Input data: numpy array of frames of a video
+        Output: classify how many frames there are between a frame and a vecctor of joints over 'length' frames.
+    """
+    frame = np.empty((1, *video.shape[1:]))
+    label = np.zeros(6)
+
+    interval = np.random.choice(6)
+
+    if interval == 0:
+        distance = 0
+    elif interval == 1:
+        distance = 1
+    elif interval == 2:
+        distance = 2
+    elif interval == 3:
+        distance = np.random.choice(np.arange(3, 4 + 1))
+    elif interval == 4:
+        distance = np.random.choice(np.arange(5, 20 + 1))
+    elif interval == 5:
+        distance = np.random.choice(np.arange(21, 150 + 1))
+    if (video.shape[0] - distance - length) < 0:
+        raise ValueError('Length of video is less than 150 frames')
+    frame_index = np.random.randint(0, video.shape[0] - distance - length)
+    joint_index = frame_index + distance
+
+    
+    frame = video[frame_index] / 255 # To normalize data
+    joint = joints[joint_index:joint_index + length] 
```

Comment:
Am I correct that this is trying to get: image_n_vec_xyz_q_dq_gripper. Will the joints parameter actually contain the xyz coordinates, joint angles, change in joint angles and gripper state?

Reply:
Yes, I do `joints = np.hstack(
(np.array(data[self.pose_name]),np.array(data['q']),np.array(data['dq']),gripper_states)) `in line 1138. 
As you suggested, I encode the pose now so I changed the data_features_to_extract to 'image_n_vec_xyz_aaxyz_nsc_q_dq_gripper'. Is the naming okay?

Reply:
sounds good!